libconnman-qt/networkmanager.cpp

```diff
@@ -827,62 +830,87 @@ void NetworkManager::propertyChanged(const QString &name, const QDBusVariant &va
+
+    // Check if it is VPN -> old is the transport as this does not understand
+    // or has not been designed VPNs in mind. So the default service from
+    // networkmanager point of view is still the transport.
+    // TODO implement a VPN support here as well
+    if (isVPN) {
+        m_defaultRouteIsVPN = true;
+        if (m_defaultRoute && m_defaultRoute != m_invalidDefaultRoute) {
+            qDebug() << "New default service is VPN, use old service " << m_defaultRoute->name();
+            return m_defaultRoute;
+        } else {
+            qDebug() << "No old default service set, select next connected";
+        }
+    } else {
+       if (m_servicesOrder.contains(path)) {
+            newDefaultRoute = m_servicesCache.value(path, nullptr);
+            if (newDefaultRoute && newDefaultRoute->connected()) {
+                qDebug() << "Selected service" << newDefaultRoute->name() << "path" << path;
+                return newDefaultRoute;
+            } else {
+                qDebug() << "Service" << (newDefaultRoute ? newDefaultRoute->name() : "NULL") << "not connected";
             }
-            line = in.readLine();
+        } else {
+            qDebug() << "No service for" << path << "select next connected";
         }
-        routeFile.close();
-    }
-    if (defaultNetDev.isNull()) {
-         QFile ipv6routeFile("/proc/net/ipv6_route");
-         if (ipv6routeFile.open(QIODevice::ReadOnly | QIODevice::Text)) {
-             QTextStream ipv6in(&ipv6routeFile);
-             QString ipv6line = ipv6in.readLine();
-             while (!ipv6line.isNull()) {
-                 QStringList ipv6lineList = ipv6line.split(QRegularExpression("\\s+"));
-                 if (ipv6lineList.size() >= 10) {
-                     if (ipv6lineList.at(0) == "00000000000000000000000000000000" &&
-                        (ipv6lineList.at(8).endsWith("3") || (ipv6lineList.at(8).endsWith("1")))) {
-                         defaultNetDev = ipv6lineList.at(9).trimmed();
-                         break;
-                     }
-                     ipv6line = ipv6in.readLine();
-                 }
-             }
-             ipv6routeFile.close();
-         }
     }
 
-    for (NetworkService *service : m_servicesCache) {
-        if (service->connected()) {
-            if (defaultNetDev == service->ethernet().value("Interface")) {
-                if (m_defaultRoute != service) {
-                    m_defaultRoute = service;
-                    Q_EMIT defaultRouteChanged(m_defaultRoute);
+    // No default route was set = VPN was connected before check use the
+    // ordered list to get the next non-VPN service that is used as
+    // transport. When VPN is set as the default service the transport is
+    // always the next non-VPN service in the list. The order is defined
+    // by the technology type when states are equal ethernet > wlan >
+    // cellular.
+    int i = 0;
+    for (const QString servicePath : m_servicesOrder) {
```

Comment:
```suggestion
    for (const QString &servicePath : m_servicesOrder) {
```

Reply:
Added in d90b034ef0128e0196b1ca097354074b1e695b79 among other changes.